File: src/ch1.go
package main

import (
   "bufio"
    "fmt"
    "os"
    "unicode"
    "regexp"
)

func check(e error) {
    if e != nil {
        panic(e)
    }
}

func main() {
	lines := readFile(os.Args[1])
	fmt.Printf("Part 1 Sum %d\n", part1(lines))
	fmt.Printf("Part 2 Sum %d\n", part2(lines))
}

func readFile(fileName string) []string {
	var lines []string

	file, err := os.Open(fileName)
	check(err)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	check(scanner.Err())
	return lines
}

func part1(lines []string) int {
	sum := 0
	for _, line := range lines {
		a := -1
		b := -1
		for _, c := range line {
			if (unicode.IsDigit(c)) {
				if (a == -1) {
					a = int(c - '0')
				} else {
					b = int(c - '0')
				}
			}
		}
		if (b == -1) {
			b = a
		}
		sum += a*10 + b
	}
	return sum
}


func part2(lines []string) int {
	numMap := map[string]int {
		"one":   1,
		"two":   2,
		"three": 3,
		"four":  4,
		"five":  5,
		"six":   6,
		"seven": 7,
		"eight": 8,
		"nine":  9,
		"0":  0,
		"1":  1,
		"2":  2,
		"3":  3,
		"4":  4,
		"5":  5,
		"6":  6,
		"7":  7,
		"8":  8,
		"9":  9,
	}

	sum := 0
	r := regexp.MustCompile(`^.*?(one|two|three|four|five|six|seven|eight|nine|[0-9]).*$`)
	r2 := regexp.MustCompile(`^.*(one|two|three|four|five|six|seven|eight|nine|[0-9]).*?$`)
	for _, line := range lines {
		v := r.FindStringSubmatch(line)[1]
		v2 := r2.FindStringSubmatch(line)[1]
		a := numMap[v]
		b := numMap[v2]
		sum += a*10 + b
	}
	return sum
}

File: src/ch1.zig
const std = @import("std");
const regex = @import("regex/src/regex.zig");

const allocator = std.heap.page_allocator;
const ArrayList = std.ArrayList;
const print = std.debug.print;

const Regex = regex.Regex;

pub fn main() !void {
    var args = std.process.args();
    _ = args.skip();

    const input_file = args.next() orelse "";
    std.debug.print("Running ch1 with {s}\n", .{input_file});

    const file = try read_file(input_file);
    const sum = part1(file);
    print("Part 1 Sum {}\n", .{sum});

    const sum2 = part2(file);
    print("Part 2 Sum {}\n", .{sum2});

    defer allocator.free(file);
}

pub fn read_file(input_file: []const u8) ![]u8 {
    var file = try std.fs.cwd().readFileAlloc(allocator, input_file, 100000);
    return file;
}

pub fn part1(file: []u8) i32 {
    var lines = std.mem.tokenize(u8, file, "\n");

    var sum: i32 = 0;
    while (lines.next()) |line| {
        // std.debug.print("{s}\n", .{line});

        var a: i32 = 0;
        for (line) |c| {
            if (std.ascii.isDigit(c)) {
                a += c - '0';
                break;
            }
        }

        var b: i32 = 0;
        var i: usize = line.len - 1;
        while (i >= 0) : (i -= 1) {
            var c = line[i];
            if (std.ascii.isDigit(c)) {
                b += c - '0';
                break;
            }
        }
        sum += a * 10 + b;
    }
    return sum;
}

const map = std.ComptimeStringMap(i32, .{
    .{ "one", 1 },
    .{ "two", 2 },
    .{ "three", 3 },
    .{ "four", 4 },
    .{ "five", 5 },
    .{ "six", 6 },
    .{ "seven", 7 },
    .{ "eight", 8 },
    .{ "nine", 9 },
    .{ "zero", 0 },
});

const numbers = "one|two|three|four|five|six|seven|eight|nine|zero";
pub fn part2(file: []u8) i32 {
    const re = try Regex.compile(allocator, "(" ++ numbers ++ "|[0-9]).*(" ++ numbers ++ "|[0-9])");
    const captures = try re.captures(file);
    print("{s} {s}", .{ captures.sliceAt(1), captures.sliceAt(1) });
    return 0;
}
